### ISSUE TITLE
Linux program termination due to pplx::details Task_impl_base waitEv rethrow exception

### ISSUE THREAD

**Issue by sujathawgp** [Microsoft/cpprestsdk#1065]
Hi,
I am using cpprest 2.10.6 on my CentOS machine (built from source at https://github.com/Microsoft/cpprestsdk/releases/tag/v2.10.6).
The program runs fine on Windows machine but fails on CentOS machine running on VM VirtualBox.

Here is the exception trace. I have added comments for few lines to make things clear.

`0x00007f68022b0207 libc.so.6!gsignal(+0x37)`

`0x00007f68022b18f8 libc.so.6!abort(+0x147)`

`0x0000000007f0e78b MyServer!abort_message()`

`0x0000000007f0eae1 MyServer!demangling_terminate_handler()()`

`0x0000000007f0e983 MyServer!std::__terminate(void (*)())()`

`0x0000000007f0ea05 MyServer!std::terminate()(+0x44) // My Server is terminating because pplx is rethrowing exception. `

`0x00007f68015d7734 libstdc++.so.6!_ZSt17rethrow_exceptionNSt15__exception_ptr13exception_ptrE(+0x73) // Libstdc++ is passing the rethrow exception to MyServer program`

`0x00007f6803327195 libMyLibrary.so!_ZN4pplx7details16_ExceptionHolder21_RethrowUserExceptionEv(+0x24) // rethrow user exception is passed to libstdc++ from my dynamic library`

`0x00007f6803327f95 libMyLibrary.so!_ZN4pplx7details15_Task_impl_base5_WaitEv(+0x2e4) // WaitEv() is rethrowing exception which is handled fine in Windows but not on CentOS`

`0x00007f6803327ffb libMyLibrary.so!_ZNK4pplx4taskISt6vectorISt10shared_ptrIN2io7swagger6client5model4NetworkEESaIS8_EEE3getEv(+0x1a) // This is getting a vector of shared pointers of object of class Network (created by using Swagger CodeGen (https://mjstealey.github.io/swagger-demo/apiclient/)`

`0x00007f6803323f11 libMyLibrary.so!_ZN5MyNetwork18NetworkManager13GetMainServerNameEPKwPwmPiPb(+0x1d0) //This is my library function which gets the main server name from my server.`

Basically what seems like the issue is that, pplx::details Task_impl_base waitEv() is rethrowing exception, which is passed to the libstdc++ which is causing the whole program to terminate instead of passing it back to the calling function. 

I tried searching answers for these questions but could not find them online.

1. Why is libstdc++ causing program to terminate when handling rethrow exception. Is there a proper way to handle it? Is there any flag to ignore it? On that regard, I have a try catch block in my program to catch any exception. But it was not helping avoid the crash.

2. Why would this be failing in Linux and not on Windows?

Can anyone shed light on this issue?

**Comment by sujathawgp** [comment]
Update: Answering my own questions and asking few more. 
1. libstdc++ is supposed to do that, hence it is normal expected behaviour from unhandled rethrown exception
2. Not sure yet. 

3. catch all phrase (catch(...)) is catching the exception, but attempting to get that exception to figure out details is not helping because current_exception() is returning null. As per the official docs http://www.cplusplus.com/reference/exception/current_exception/ 

> If no exception is being handled, the function returns a null-pointer value. 

Here is where it does not seem to make sense. Shouldn't it be non-null when execution of catch-all happens, because it caught some kind of exception? Anyone know why it might be returning null, in this specific case?
 
  ```
 catch (...)
        { 
      auto eptr = std::current_exception();
        try
        {
            if (eptr)
            {
                std::rethrow_exception(eptr);
            }
            else
            {
               //This line gets executed because eptr is null. Here is where I am confused. Shouldn't it be non null when execution of catch all happens, because it caught some kind of exception?
                throw MyException(999, string("null eptr"));//MyException is derived class of std::exception
            }
        }
        catch (const std::exception& e)
        {
           //log the message:e.what() here
        }
    }


```

4. Anyone know of any sample project that I can reference to, that works on Linux, and uses pplx::task of cpprest?
